server/src-lib/Hasura/GraphQL/Execute.hs

```diff
@@ -225,207 +217,103 @@ getResolvedExecPlan
-  -> QueryActionExecuter
-  -> VQ.ObjectSelectionSet
-  -> m (EL.LiveQueryPlan, Maybe EL.ReusableLiveQueryPlan)
-getSubsOp env logger pgExecCtx gCtx sqlGenCtx userInfo queryReusability actionExecuter =
-  runE logger gCtx sqlGenCtx userInfo .
-  EL.buildLiveQueryPlan env pgExecCtx queryReusability actionExecuter
+      let takeFragment = \case G.ExecutableDefinitionFragment f -> Just f; _ -> Nothing
+          fragments =
+            mapMaybe takeFragment $ unGQLExecDoc $ _grQuery reqParsed
+      (gCtx, queryParts) <- getExecPlanPartial userInfo sc queryType reqParsed
+      case queryParts of
+        G.TypedOperationDefinition G.OperationTypeQuery _ varDefs _ selSet -> do
+          -- (Here the above fragment inlining is actually executed.)
+          inlinedSelSet <- EI.inlineSelectionSet fragments selSet
+          -- (unpreparedQueries, _) <-
+          --   E.parseGraphQLQuery gCtx varDefs
+          (execPlan,asts) {-, plan-} <-
+            EQ.convertQuerySelSet env logger gCtx userInfo httpManager reqHeaders inlinedSelSet varDefs (_grVariables reqUnparsed)
+          -- See Note [Temporarily disabling query plan caching]
+          -- traverse_ (addPlanToCache . EP.RPQuery) plan
+          return $ QueryExecutionPlan execPlan asts
+        G.TypedOperationDefinition G.OperationTypeMutation _ varDefs _ selSet -> do
+          -- (Here the above fragment inlining is actually executed.)
+          inlinedSelSet <- EI.inlineSelectionSet fragments selSet
+          queryTx <- EM.convertMutationSelectionSet env logger gCtx sqlGenCtx userInfo httpManager reqHeaders
+                     inlinedSelSet varDefs (_grVariables reqUnparsed)
+          -- See Note [Temporarily disabling query plan caching]
+          -- traverse_ (addPlanToCache . EP.RPQuery) plan
+          return $ MutationExecutionPlan queryTx
+        G.TypedOperationDefinition G.OperationTypeSubscription _ varDefs directives selSet -> do
+          -- (Here the above fragment inlining is actually executed.)
+          inlinedSelSet <- EI.inlineSelectionSet fragments selSet
+          -- Parse as query to check correctness
+          (unpreparedAST, _reusability) <-
+            EQ.parseGraphQLQuery gCtx varDefs (_grVariables reqUnparsed) inlinedSelSet
+          -- A subscription should have exactly one root field
+          -- As an internal testing feature, we support subscribing to multiple
+          -- root fields in a subcription. First, we check if the corresponding directive
+          -- (@_multiple_top_level_fields) is set.
+          case NE.nonEmpty inlinedSelSet of
+            Nothing -> throw500 "empty selset for subscription"
+            Just (_ :| rst) ->
+              let multipleAllowed =
+                    G.Directive $$(G.litName "_multiple_top_level_fields") mempty `elem` directives
+              in
+              unless (multipleAllowed || null rst) $
+                throw400 ValidationFailed "subscriptions must select one top level field"
```

Comment:
There's no point in converting to `NonEmpty`? The same check could be written as:
```
case inlinedSelSet of
  []      -> throw500 ...
  (_:rst) ->
```

Reply:
Oh yes, your suggestion makes more sense! I'll make the change!

Reply:
Done